src/main/java/com/typesafe/netty/http/pipelining/HttpPipeliningHandler.java

```diff
@@ -1,110 +1,208 @@
+ * inFlightRequestsLowWatermark and inFlightRequestsHighWatermark. When the high watermark is exceeded, the handler will
+ * push back on the client. When the low watermark is reached, the handler will start reading again.
  */
-public class HttpPipeliningHandler extends SimpleChannelHandler {
+public class HttpPipeliningHandler extends ChannelDuplexHandler {
 
-    public static final int INITIAL_EVENTS_HELD = 3;
-    public static final int MAX_EVENTS_HELD = 10000;
+    /**
+     * The sequence of received HTTP requests.
+     */
+    volatile private int receiveSequence = 0;
 
-    private final int maxEventsHeld;
+    /**
+     * The currently sending sequence of HTTP requests.
+     */
+    volatile private int currentlySendingSequence = 1;
 
-    private int sequence;
-    private int nextRequiredSequence;
-    private int nextRequiredSubsequence;
+    /**
+     * Whether the high watermark has been exceeded.
+     */
+    volatile private boolean highWatermarkExceeded = false;
 
-    private final Queue<OrderedDownstreamChannelEvent> holdingQueue;
+    /**
+     * The number of requests in flight
+     */
+    private final AtomicInteger inFlight = new AtomicInteger();
 
-    public HttpPipeliningHandler() {
-        this(MAX_EVENTS_HELD);
+    /**
+     * A write message with a promise of when it's written.
+     */
+    private static class WriteMessage {
+        /**
+         * The written message.
+         */
+        final SequencedOutboundMessage message;
+
+        /**
+         * The future that is redeemed once the message is written.
+         */
+        final ChannelPromise promise;
+
+        public WriteMessage(SequencedOutboundMessage message, ChannelPromise promise) {
+            this.message = message;
+            this.promise = promise;
+        }
     }
 
     /**
-     * @param maxEventsHeld the maximum number of channel events that will be retained prior to aborting the channel
-     *                      connection. This is required as events cannot queue up indefintely; we would run out of
-     *                      memory if this was the case.
+     * The buffered events, by sequence
      */
-    public HttpPipeliningHandler(final int maxEventsHeld) {
-        this.maxEventsHeld = maxEventsHeld;
-
-        holdingQueue = new PriorityQueue<OrderedDownstreamChannelEvent>(INITIAL_EVENTS_HELD, new Comparator<OrderedDownstreamChannelEvent>() {
-            @Override
-            public int compare(OrderedDownstreamChannelEvent o1, OrderedDownstreamChannelEvent o2) {
-                final int delta = o1.getOrderedUpstreamMessageEvent().getSequence() - o2.getOrderedUpstreamMessageEvent().getSequence();
-                if (delta == 0) {
-                    return o1.getSubsequence() - o2.getSubsequence();
-                } else {
-                    return delta;
-                }
-            }
-        });
+    private final Map<Integer, List<WriteMessage>> bufferedEvents = new HashMap<>();
```

Comment:
how about this?
https://github.com/netty/netty/blob/4.0/common/src/main/java/io/netty/util/collection/IntObjectHashMap.java